packages/compiler-core/src/transforms/transformElement.ts

```diff
@@ -747,7 +748,7 @@ export function buildProps(
     }
   }
   if (
-    !shouldUseBlock &&
+    !(shouldUseBlock && !hasRef) &&
```

Comment:
I'm unclear why `hasRef` has been singled out for special treatment. Two lines down from here, we check for `hasRef || hasVnodeHook || runtimeDirectives.length > 0`. Don't we have the same problem for `hasVnodeHook` and `runtimeDirectives.length > 0`?

* VNode hook example: [SFC Playground](https://play.vuejs.org/#eNp9Uk1PwkAQ/SvjXoAEgYQbKQQ1HPSgRomnvUA7rQvb3WY/Kobw351taanG0Esn895L33vTI7srilHpkc1YZGMjCgcWnS9AblQ258xZzhZcibzQxsERDKZwgtToHHok63HFVayVdSDsh7BiKxHmgdV3xuMgwKlXsRNaQa69cpj0B3DkCiDItMSR1Fm/d8Z6pDj9Enl1TdaiZ2E0rkOQ5chhXsiNQ5oBokSUUN6KlDK1TqtoQE9A6wmIlGpDrH5ZToYgVIKHyYDeMOWs4Swp+uz8ZaKepz94640Y7XzhzPb4HcxUH2jWi2hMVmrHl2nrnaMqlrEU8b4bgKq+6aZZ6yyTGI1rQShh3GmBDemcVF8qstHOakU3rxrlLNZ5ISSalyJ0Tief1V0HbCOl/nqqduGkw2Yff2K8/2e/s4ew4+zVoEVTkq8WcxuToavh1fszHmhuwVwnPqS4Ar4hnd4HjzXt3quEbHd4ldvH6m8VKlvb1cGhsk2oYDQwTxWfMzrTw5XoF7vT0bTS0U/GTj/VsBFL)
* Runtime directive example: [SFC Playground](https://play.vuejs.org/#eNp9UstuwjAQ/JWtL4BEAYkbCqgvDu2hrVrUky+QbFKDY0d+pFSIf+86IY9DhRQla8/sZmZ3T+y+KCalR7ZgkY2NKBxYdL4AuVXZkjNnOVtxJfJCGwcnMJjCGVKjcxhQ2oArrmKtrANhv4QVO4mwDKyhMx5HHVw+CUPIiSuAXHvlMBmO6iNAoGiJE6mz4eCCDigZ4DwOb6+up7T4JYkreqJpbYjkRw7zQm4dUgwQJaKE8lak5K9VXdkMhQNaR0CkVBtiDctyNgahEjzORvSFOWcdJxGmOSwO+BuqVsyGs4qmVLP+dRftvHNawV0sRXzoK6Eu3fRlbXSWSYymdUJwM+3ZYWOaETUjFdlkb7WiQVYd4izWeSEkmrfCCWoWZ4umd5xtpdQ/L9VdmFPV5CrnG+PDP/d7ewx3nL0btGhK0tVibmsydDW8/nzFI8UtmOvEBxdXwA+kQfqgsaY9eJWQ7B6vUvtcraBQ2caujw6VbUwFod2qcEZr+XjFeid3Ppk328LOf5wm/pM=)

In both cases, the `key` seems to prevent `unmounted` being triggered.